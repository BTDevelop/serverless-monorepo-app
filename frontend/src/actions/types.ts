const _REQUESTED = '_REQUESTED';
const _FULFILLED = '_FULFILLED';
const _REJECTED = '_REJECTED';

export const LOGIN_REQUESTED = 'LOGIN' + _REQUESTED;
export const LOGIN_FULFILLED = 'LOGIN' + _FULFILLED;
export const LOGIN_REJECTED = 'LOGIN' + _REJECTED;

export const LOGOUT_REQUESTED = 'LOGOUT' + _REQUESTED;
export const LOGOUT_FULFILLED = 'LOGOUT' + _FULFILLED;
export const LOGOUT_REJECTED = 'LOGOUT' + _REJECTED;

export const SYNC_AUTH_STATE_REQUESTED = 'SYNC_AUTH_STATE' + _REQUESTED;
export const SYNC_AUTH_STATE_FULFILLED = 'SYNC_AUTH_STATE' + _FULFILLED;
export const SYNC_AUTH_STATE_REJECTED = 'SYNC_AUTH_STATE' + _REJECTED;

export const API_REQUESTED = 'API' + _REQUESTED;
export const API_FULFILLED = 'API' + _FULFILLED;
export const API_REJECTED = 'API' + _REJECTED;
