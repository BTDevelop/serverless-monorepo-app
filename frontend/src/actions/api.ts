import { createAction } from 'redux-actions';
import * as types from './types';

export const apiMetaTypes = {
  dbCreate: 'dbCreate',
  email: 'email',
  file: 'file',
  private: 'private',
  public: 'public',
};

export const apiRequested = createAction(
  types.API_REQUESTED,
  undefined,
  (_, type) => ({ type }),
);
export const apiFulfilled = createAction(
  types.API_FULFILLED,
  undefined,
  (_, type) => ({ type }),
);
export const apiRejected = createAction(
  types.API_REJECTED,
  undefined,
  (_, type) => ({ type }),
);

export const emailApiRequested = (toAddress: string) =>
  apiRequested(toAddress, apiMetaTypes.email);

export const privateApiRequested = () =>
  apiRequested(undefined, apiMetaTypes.private);

export const publicApiRequested = () =>
  apiRequested(undefined, apiMetaTypes.public);

export const fileApiRequested = (fileUrl: string, key: string) =>
  apiRequested({ fileUrl, key }, apiMetaTypes.file);

export const dbCreateApiRequested = (text: string) =>
  apiRequested(text, apiMetaTypes.dbCreate);
