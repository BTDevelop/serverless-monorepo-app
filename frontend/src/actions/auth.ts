import { createAction } from 'redux-actions';
import * as types from './types';

export const loginRequested = createAction(types.LOGIN_REQUESTED);
export const loginFulfilled = createAction(types.LOGIN_FULFILLED);
export const loginRejected = createAction(types.LOGIN_REJECTED);

export const logoutRequested = createAction(types.LOGOUT_REQUESTED);
export const logoutFulfilled = createAction(types.LOGOUT_FULFILLED);
export const logoutRejected = createAction(types.LOGOUT_REJECTED);

export const syncAuthStateRequested = createAction(
  types.SYNC_AUTH_STATE_REQUESTED,
);
export const syncAuthStateFulfilled = createAction(
  types.SYNC_AUTH_STATE_FULFILLED,
);
export const syncAuthStateRejected = createAction(
  types.SYNC_AUTH_STATE_REJECTED,
);
