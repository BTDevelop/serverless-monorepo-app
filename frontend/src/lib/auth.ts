import { Auth0DecodedHash, WebAuth } from 'auth0-js';

const auth = new WebAuth({
  audience: process.env.REACT_APP_AUTH0_AUDIENCE,
  clientID: process.env.REACT_APP_AUTH0_CLIENT_ID || '',
  domain: process.env.REACT_APP_AUTH0_DOMAIM || '',
  redirectUri: process.env.REACT_APP_AUTH0_REDIRECT_URI,
  responseType: 'token id_token',
  scope: 'openid profile email',
});

const setSession = (authResult: Auth0DecodedHash) => {
  const { expiresIn = 0, accessToken = '', idToken = '' } = authResult;
  // Set the time that the access token will expire at
  const expiresAt = JSON.stringify(expiresIn * 1000 + new Date().getTime());
  localStorage.setItem('access_token', accessToken);
  localStorage.setItem('id_token', idToken);
  localStorage.setItem('expires_at', expiresAt);
};

const handleAuthentication = async () => {
  const promise: Promise<Auth0DecodedHash> = new Promise((resolve, reject) => {
    auth.parseHash((err, authResult) => {
      const result = authResult || {};
      if (err) {
        reject(err);
      } else {
        setSession(result);
        resolve(result);
      }
    });
  });
  return await promise;
};

const login = (state: string) => auth.authorize({ state });

const logout = () => {
  localStorage.removeItem('access_token');
  localStorage.removeItem('id_token');
  localStorage.removeItem('expires_at');
};

const isAuthenticated = () => {
  // Check whether the current time is past the
  // access token's expiry time
  const expiresAt = JSON.parse(
    localStorage.getItem('expires_at') || JSON.stringify(0),
  );
  return new Date().getTime() < expiresAt;
};

export { handleAuthentication, login, logout, isAuthenticated };
