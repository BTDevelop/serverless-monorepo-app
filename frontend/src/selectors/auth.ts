import { IState } from '../types/redux';

export const isAuthenticated = (state: IState) => state.auth.authenticated;
export const isNotAuthenticated = (state: IState) => !state.auth.authenticated;
export const isInitialized = (state: IState) => state.auth.initialized;
