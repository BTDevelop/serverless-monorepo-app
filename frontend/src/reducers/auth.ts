import { Action, handleActions } from 'redux-actions';
import * as types from '../actions/types';
import { IAuthState } from '../types/redux';

const defaultState: IAuthState = {
  authenticated: false,
  error: null,
  inProgress: false,
  initialized: false,
};

const reducer = handleActions<IAuthState, any>(
  {
    [types.LOGIN_REQUESTED]: state => {
      return { ...state, inProgress: true };
    },
    [types.LOGIN_FULFILLED]: state => {
      return { ...state, inProgress: false, authenticated: true, error: null };
    },
    [types.LOGIN_REJECTED]: (state, action) => {
      return {
        ...state,
        authenticated: false,
        error: action.payload,
        inProgress: false,
      };
    },
    [types.LOGOUT_REQUESTED]: state => {
      return { ...state, inProgress: true };
    },
    [types.LOGOUT_FULFILLED]: state => {
      return { ...state, inProgress: false, authenticated: false, error: null };
    },
    [types.LOGOUT_REJECTED]: (state, action: Action<Error>) => {
      return {
        ...state,
        authenticated: false,
        error: action.payload,
        inProgress: false,
      };
    },
    [types.SYNC_AUTH_STATE_REQUESTED]: state => {
      return { ...state, inProgress: true };
    },
    [types.SYNC_AUTH_STATE_FULFILLED]: (_, action: Action<boolean>) => {
      return {
        authenticated: !!action.payload,
        error: null,
        inProgress: false,
        initialized: true,
      };
    },
    [types.SYNC_AUTH_STATE_REJECTED]: (_, action: Action<Error>) => {
      return {
        authenticated: false,
        error: action.payload,
        inProgress: false,
        initialized: true,
      };
    },
  },
  defaultState,
);

export default reducer;
