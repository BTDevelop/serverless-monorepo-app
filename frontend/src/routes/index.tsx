import { push } from 'connected-react-router';
import * as React from 'react';
import * as Loadable from 'react-loadable';
import { Route } from 'react-router';
import locationHelperBuilder from 'redux-auth-wrapper/history4/locationHelper';
import { connectedReduxRedirect } from 'redux-auth-wrapper/history4/redirect';
import Loading from '../components/Loading';
import { ComponentLoading } from '../components/Loading';
import { isAuthenticated, isNotAuthenticated } from '../selectors/auth';

export const paths = {
  callback: '/callback',
  home: '/',
  login: '/login',
  private: '/private',
  public: '/public',
};

export const userIsAuthenticated = connectedReduxRedirect({
  AuthenticatingComponent: Loading,
  // If selector is true, wrapper will not redirect
  // For example let's check that state contains user data
  authenticatedSelector: isAuthenticated,
  redirectAction: push,
  // The url to redirect user to if they fail
  redirectPath: paths.login,
  // A nice display name for this check
  wrapperDisplayName: 'UserIsAuthenticated',
});

const locationHelper = locationHelperBuilder({});

export const userIsNotAuthenticated = connectedReduxRedirect({
  // This prevents us from adding the query parameter when we send the user away from the login page
  allowRedirectBack: false,
  // If selector is true, wrapper will not redirect
  // So if there is no user data, then we show the page
  authenticatedSelector: isNotAuthenticated,
  redirectAction: push,
  // This sends the user either to the query param route if we have one, or to the landing page if none is specified and the user is already logged in
  redirectPath: (_, ownProps) =>
    locationHelper.getRedirectQueryParam(ownProps) || paths.home,
  // A nice display name for this check
  wrapperDisplayName: 'UserIsNotAuthenticated',
});

export const routes = {
  Callback: (
    <Route
      path={paths.callback}
      component={Loadable({
        loader: () => import('../components/Callback'),
        loading: ComponentLoading,
      })}
    />
  ),
  Home: (
    <Route
      exact={true}
      path={paths.home}
      component={Loadable({
        loader: () => import('../components/Home'),
        loading: ComponentLoading,
      })}
    />
  ),
  Login: (
    <Route
      path={paths.login}
      component={userIsNotAuthenticated(
        Loadable({
          loader: () => import('../components/Login'),
          loading: ComponentLoading,
        }),
      )}
    />
  ),
  NotFound: (
    <Route
      component={Loadable({
        loader: () => import('../components/NotFound'),
        loading: ComponentLoading,
      })}
    />
  ),
  Private: (
    <Route
      path={paths.private}
      component={userIsAuthenticated(
        Loadable({
          loader: () => import('../components/Private'),
          loading: ComponentLoading,
        }),
      )}
    />
  ),
  Public: (
    <Route
      path={paths.public}
      component={Loadable({
        loader: () => import('../components/Public'),
        loading: ComponentLoading,
      })}
    />
  ),
};
