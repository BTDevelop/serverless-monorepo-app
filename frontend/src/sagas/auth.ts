import { call, put, select, takeEvery } from 'redux-saga/effects';
import * as types from '../actions/types';
import { isAuthenticated, login, logout } from '../lib/auth';
import {
  logoutFulfilled,
  logoutRejected,
  syncAuthStateFulfilled,
  syncAuthStateRejected,
} from './../actions/auth';
import { redirectRouteSelector } from './../selectors/router';

export function* handleLogin() {
  const redirectRoute = yield select(redirectRouteSelector);
  yield call(login, redirectRoute);
}

export function* loginSaga() {
  yield takeEvery(types.LOGIN_REQUESTED, handleLogin);
}

export function* handleLogout() {
  try {
    yield call(logout);
    yield put(logoutFulfilled());
  } catch (err) {
    yield put(logoutRejected(err));
  }
}

export function* logoutSaga() {
  yield takeEvery(types.LOGOUT_REQUESTED, handleLogout);
}

export function* handleAuthState() {
  try {
    const authenticated = yield call(isAuthenticated);
    yield put(syncAuthStateFulfilled(authenticated));
  } catch (err) {
    yield put(syncAuthStateRejected(err));
  }
}

export function* authStateSaga() {
  yield takeEvery(types.SYNC_AUTH_STATE_REQUESTED, handleAuthState);
}

const sagas = [loginSaga(), logoutSaga(), authStateSaga()];

export default sagas;
