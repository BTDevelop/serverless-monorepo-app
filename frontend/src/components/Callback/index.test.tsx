import { replace } from 'connected-react-router';
import { shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { loginFulfilled, loginRejected } from '../../actions/auth';
import { paths } from '../../routes';
import ConnectedCallback, { Callback } from './';

jest.mock('../../lib/auth');
jest.mock('../Loading', () => {
  return {
    ComponentLoading: 'component-loading-component',
    default: 'loading-component',
  };
});

const middlewares: any[] = [];
const mockStore = configureStore(middlewares);

const props = {
  loginFulfilled: jest.fn(),
  loginRejected: jest.fn(),
  replace: jest.fn(),
};

describe('Callback Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should render correctly', () => {
    const tree = renderer.create(<Callback {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('handles authentication on mount with auth state', async () => {
    const { handleAuthentication } = require('../../lib/auth');

    const wrapper = shallow(<Callback {...props} />, {
      disableLifecycleMethods: true,
    });

    const authResult = { state: 'state' };
    handleAuthentication.mockReturnValueOnce(Promise.resolve(authResult));

    // @ts-ignore access to undefined
    await wrapper.instance().componentDidMount();

    expect(handleAuthentication).toHaveBeenCalledTimes(1);
    expect(props.loginFulfilled).toHaveBeenCalledTimes(1);
    expect(props.loginFulfilled).toHaveBeenCalledWith(authResult);
    expect(props.replace).toHaveBeenCalledTimes(1);
    expect(props.replace).toHaveBeenCalledWith(authResult.state);
  });

  test('handles authentication on mount no auth state', async () => {
    const { handleAuthentication } = require('../../lib/auth');

    const wrapper = shallow(<Callback {...props} />, {
      disableLifecycleMethods: true,
    });

    const authResult = {};
    handleAuthentication.mockReturnValueOnce(Promise.resolve(authResult));

    // @ts-ignore access to undefined
    await wrapper.instance().componentDidMount();

    expect(props.replace).toHaveBeenCalledWith('/public');
  });

  test('handles authentication on mount with error', async () => {
    const { handleAuthentication } = require('../../lib/auth');

    const wrapper = shallow(<Callback {...props} />, {
      disableLifecycleMethods: true,
    });

    const error = new Error('error');
    handleAuthentication.mockReturnValueOnce(Promise.reject(error));

    // @ts-ignore access to undefined
    await wrapper.instance().componentDidMount();

    expect(props.loginRejected).toHaveBeenCalledTimes(1);
    expect(props.loginRejected).toHaveBeenCalledWith(error);
    expect(props.replace).toHaveBeenCalledTimes(1);
    expect(props.replace).toHaveBeenCalledWith(paths.login);
  });

  test('should connect component correctly', () => {
    const store = mockStore();

    // @ts-ignore wrong store type
    const wrapper = shallow(<ConnectedCallback store={store} />);

    wrapper.props().loginFulfilled();
    wrapper.props().loginRejected();
    wrapper.props().replace('/');

    expect(store.getActions()).toHaveLength(3);
    expect(store.getActions()[0]).toEqual(loginFulfilled());
    expect(store.getActions()[1]).toEqual(loginRejected());
    expect(store.getActions()[2]).toEqual(replace(paths.home));
  });
});
