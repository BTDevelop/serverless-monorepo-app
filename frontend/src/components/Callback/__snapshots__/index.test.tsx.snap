// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Callback Component should render correctly 1`] = `<loading-component />`;
